pants_jupyter_plugin/plugin.py

```diff
@@ -0,0 +1,403 @@
+        raise self.SubprocessFailure(
+          f'command `{cmd}` failed with exit code {return_code}',
+          return_code=return_code
+        )
+
+    with self._accordion_widget(title, collapsed=False) as (expand, collapse, set_output_glyph):
+      self._display_line(f'$ {cmd}\n')
+      is_complete = asyncio.Event()
+
+      try:
+        run_async(
+          async_exec(self._display_line, cmd, title, is_complete),
+          spin_driver(set_output_glyph, is_complete)
+        )
+        resulting_binary = self._extract_resulting_binary(work_dir, extension)
+        self._display_line(f'\nSuccessfully built {resulting_binary}')
+
+        set_output_glyph(SUCCESS_GLYPH)
+        collapse()
+        return resulting_binary
+      except self.SubprocessFailure:
+        try:
+          set_output_glyph(FAIL_GLYPH)
+          expand()
+          self._display_line('\n\n')
+        finally:
+          raise
+
+  def _run_pex(self, requirements: str) -> pathlib.PosixPath:
+    """Runs pex with widget UI display."""
+    with temporary_dir(cleanup=False) as tmp_dir:
+      tmp_path = pathlib.PosixPath(tmp_dir)
+      output_pex = tmp_path.joinpath('requirements.pex')
+      title = f'[Resolve] {requirements}'
+      safe_requirements = ' '.join(shlex.quote(r) for r in shlex.split(requirements))
+      # TODO: Add support for toggling `--no-pypi` and find-links/index configs.
+      cmd = f'pex -vv -o "{output_pex}" {safe_requirements}'
+      return self._stream_binary_build_with_output(cmd, title, tmp_path, extension='pex')
+
+  def _run_pants(
+    self,
+    pants_repo: pathlib.PosixPath,
+    pants_target: str,
+    extension: str
+  ) -> pathlib.PosixPath:
+    """Runs pants with widget UI display."""
+
+    # Version check for pants v1 vs v2 flags/behavior.
+    is_pants_v1 = pants_repo.joinpath('pants.ini').exists()
+    if is_pants_v1:
+      goal_name = 'binary'
+      tmp_root = None
+    else:
+      goal_name = 'package'
+      # N.B. pants v2 doesn't support `--pants-distdir` outside of the build root.
+      tmp_root = pants_repo.joinpath('dist')
+      # N.B. The dist dir must exist for temporary_dir.
+      tmp_root.mkdir(exist_ok=True)
+
+    with temporary_dir(root_dir=tmp_root, cleanup=False) as tmp_dir:
+      tmp_path = pathlib.PosixPath(tmp_dir)
+      title = f'[Build] ./pants {goal_name} {pants_target}'
+      cmd = f'cd {pants_repo} && ./pants --pants-distdir="{tmp_path}" {goal_name} {pants_target}'
+      return self._stream_binary_build_with_output(cmd, title, tmp_path, extension=extension)
+
+  def _bootstrap_pex(self, pex_path: pathlib.PosixPath):
+    """Bootstraps a pex with widget UI display."""
+    title = f'[Bootstrap] {pex_path.name}'
+    with self._accordion_widget(title) as (expand, collapse, set_output_glyph):
+      try:
+        with environment_as(PEX_VERBOSE='2'):
+          # Scrub the environment.
+          _scrub_import_environment(self._ORIGINATING_SYS_MODULES_KEYS, self._display_line)
```

Comment:
I'd like to follow up with a PR that proposes this change from using the Pex API to using the PEX_TOOLS cli to install the PEX into a tmp? dir (Basically use the venv tool minus setting up the venv; i.e.: just populate a tmp dir site-packages) yielding a single sys.path entry to scrub / add. It should simplify and robustify this code against Pex changes.